datalad/support/gitrepo.py

```diff
@@ -2274,16 +2274,21 @@ def cherry_pick(self, commit):
         self._git_custom_command("", ["git", "cherry-pick", commit],
                                  check_fake_dates=True)
 
-    # run() needs this ATM, but should eventually be RF'ed to a
-    # status(recursive=True) call
     @property
     def dirty(self):
-        return len([
-            p for p, props in iteritems(self.status(
-                untracked='all', eval_submodule_state='full'))
-            if props.get('state', None) != 'clean' and
-            # -core ignores empty untracked directories, so shall we
-            not (p.is_dir() and len(list(p.iterdir())) == 0)]) > 0
+        """Is the repository dirty?
+
+        Note: This provides a quick answer when you simply want to know if
+        there are any untracked changes or modifications in this repository or
+        its submodules. For finer-grained control and more detailed reporting,
+        use status() instead.
+        """
+        stdout, _ = self._git_custom_command(
+            [],
+            ["git", "status", "--porcelain",
+             # Ensure the result isn't influenced by diff.ignoreSubmodules.
+             "--ignore-submodules=none"])
+        return bool(stdout.strip())
```

Comment:
I am curious on how does `git status` work.  
E.g. in a big repo with lots of dirty things -- wouldn't we then first get the entire output (might take awhile?) first to be stripped and then checked for bool value?  i wonder if there is an easy way to get only the first line from `status` output without too much dance with our Runner ...  Or that all shouldn't matter much?

NB original implementation should have used `return any(p for p, ...)` construct to quit without analyzing the entire status list...

Reply:
I don't know I way to do the status check and not get all the output.  Suggestions welcome.

That being said, it is more efficient than an ls-{tree,file} approach that lists all content in the repo.

Reply:
yeah, it should be.
I was googling up, apparently the same question ("quick git dirty check") comes often in shell prompt scenarios.  In  https://github.com/sindresorhus/pure/issues/115#issuecomment-107365697 spotted them using `-unormal` with their `status --porcelain`.  For me `git status --help` says

```
       -u[<mode>], --untracked-files[=<mode>]
           Show untracked files.

           The mode parameter is used to specify the handling of untracked files. It is optional: it defaults
           to all, and if specified, it must be stuck to the option (e.g.  -uno, but not -u no).

           The possible options are:
           •   no - Show no untracked files.
           •   normal - Shows untracked files and directories.
           •   all - Also shows individual files in untracked directories.
```
so I expected it to show `all` files, but for me it functions as `-unormal`:

```shell
$> time git status --porcelain | head
?? .datalad/                         
?? .git-aside/
?? .gitattributes
?? logs/
?? sub-000001_ses-01_run-01_task-efnback.feat/
?? sub-000001_ses-01_run-01_task-efnback.fsf
?? sub-000001_ses-01_run-02_task-efnback.feat/
?? sub-000001_ses-01_run-02_task-efnback.fsf
?? sub-000001_ses-02_run-01_task-efnback.feat/
?? sub-000001_ses-02_run-01_task-efnback.fsf
git status --porcelain  0.01s user 0.02s system 98% cpu 0.028 total
head  0.00s user 0.00s system 5% cpu 0.027 total

$> time git status --porcelain -uall | head
?? .datalad/.gitattributes
?? .datalad/config
?? .git-aside/HEAD
?? .git-aside/annex/index
?? .git-aside/annex/index.lck
?? .git-aside/annex/journal.lck
```
did I misread or the man page is wrong or I have some settings in git config ? ;)  (git 2.21.0.593.g511ec345e18)

anyways -- may be it would be wortwhile to hardcode `-unormal` to avoid possible side-effects from configs etc?

Reply:
> might take awhile?

Here are timings for ~30000 modified or untracked files across 3 directories,

<details>
<summary>script to generate</summary>

```sh
#!/bin/sh

set -eu

make_tree () {
    parallel -i sh -c "dd if=/dev/urandom of={}.txt count=1 bs=1024 2>>out" -- \
             $(seq 1 10000)
}

random_tree () {
    subdir=$1
    mkdir -p $subdir
    (cd $subdir && make_tree)
}

cd $(mktemp -dt dl-status-XXXXXXX)
git init

random_tree dir-a
random_tree dir-b

git add .
git commit -minitial

# Modify existing.
random_tree dir-a
random_tree dir-b
# Add untracked.
random_tree dir-c
```

</details>

Floor timing for `git status --porcelain`:

```
% time git status --porcelain >/dev/null
git status --porcelain >/dev/null  0.19s user 0.12s system 108% cpu 0.289 total
```

GitRepo.dirty on this PR:

```
% git -C ~/src/python/datalad rev-parse HEAD
e92689cd02b4a3c41359bcb7e46c80970b7bb395
% python -m timeit -s "import datalad.api as dl" "dl.Dataset('.').repo.dirty"
10 loops, best of 3: 281 msec per loop
```

GitRepo.dirty on master:

```
% git -C ~/src/python/datalad rev-parse HEAD
e83d571b0c551a2b8c8c4c1a526365df89f07965
% python -m timeit -s "import datalad.api as dl" "dl.Dataset('.').repo.dirty"
10 loops, best of 3: 2.02 sec per loo
```

GitRepo.dirty on 0.11.x:

```
% git -C ~/src/python/datalad rev-parse HEAD
78c72ec23cdd659580166de6b8313bab68decddd
% python -m timeit -s "import datalad.api as dl" "dl.Dataset('.').repo.dirty"
10 loops, best of 3: 312 msec per loop
```

I think this test repo is very much the extreme of what we could expect for number of modified files (but please modify and test on your end), so I'd say we're OK with this PR's current approach until we actually observe an issue to address.

Reply:
> did I misread or the man page is wrong or I have some settings in git config ? ;) (git 2.21.0.593.g511ec345e18)

My understanding is that if `-u` is given without an argument, it defaults to `all`, but if `-u` isn't given at all, then the behavior matches normal (mentioned just below the text you quote and in `git-config`s description of `status.showUntrackedFiles`.

> anyways -- may be it would be wortwhile to hardcode `-unormal` to avoid possible side-effects from configs etc?

Oh, good point!  I actually didn't know that `status.showUntrackedFiles` existed, though even if it didn't it is best to be explicit rather than relying on the default behavior being `normal` (as I already did with `--ignore-submodules=none`).  Will update.